daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,177 @@
+
+  /**
+   * Method to be called whenever any element that is a Term is encountered.
+   * This applies to Sequence, Choice, GroupRef, ElementBase, etc.
+   *
+   * It is highly recommended that, when implementing this method, you pattern match
+   * some of these different sub-types at some point to handle each accordingly.
+   *
+   * @param termElement the term element
+   */
+  def onTermBegin(termElement: TermView): Unit
+
+  /**
+   * Method to be called when a Term element has finished processing
+   *
+   * @param termElement the term element
+   */
+  def onTermEnd(termElement: TermView): Unit
+
+  /**
+   * Method to be called when either a Simple or Complex Type element has been encountered.
+   * This is just the element for <simpleType> or <complexType>, so no implementation is necessary if you
+   * do not care about these wrapper portion; its children will be walked automatically.
+   *
+   * @param typeElement either a Simple or Complex type
+   */
+  def onTypeBegin(typeElement: TypeView): Unit
+
+  /**
+   * Method to be called when either a Simple or Complex Type element has been finished
+   * processing.  See onTypeBegin method description
+   *
+   * @param typeElement either a Simple or Complex type
+   */
+  def onTypeEnd(typeElement: TypeView): Unit
+
+  /**
+   * Starts a DSOM walk from the Root object.  By default, the walk is completed recursively in the helper
+   * method below, but a custom walk can be defined by overriding
+   * this method instead and calling local event handlers.
+   * @param schemaSetRoot The root element of the DFDL Schema.  This will be the starting point of the traversal
+   */
+  def walkFromRoot(schemaSetRoot: RootView): Unit = {
+    onWalkBegin(schemaSetRoot)
+    walkerHelper(schemaSetRoot)
+    // onWalkEnd will return the value returned by this method!
+    onWalkEnd(schemaSetRoot)
+  }
+
+  /**
+   * Local helper method used to recursively walk the DSOM.
+   * If a non-recursive walk or a more specific recursive walk is desired, one
+   * can override this method, or simply not use it and instead override walkFromRoot
+   * and have that overridden method reference custom helper methods.
+   * @param termView the current Term element to be pattern matched
+   */
+  protected def walkerHelper(termView: TermView): Unit = {
+    onTermBegin(termView)
+    termView match {
+      case element: ElementBaseView =>
+        if (element.isComplexType) {
+          onTypeBegin(element.complexType)
+          walkerHelper(element.complexType.group)
+          onTypeEnd(element.complexType)
+        } else {
+          onTypeBegin(element.simpleType)
+          onTypeEnd(element.simpleType)
+        }
+      case groupRef: GroupRefView =>
+        if (!groupRef.isHidden) {
+          groupRef.groupMembers.foreach(walkerHelper)
+        }
+      case modelGroup: ModelGroupView =>
+        modelGroup.groupMembers.foreach(walkerHelper)
+      case _ =>
```

Comment:
Does thid default case ever get hit? I think this is impossible?

Reply:
As the View structure currently stands, the default case will not ever get hit; removed this case clause in the latest revision.